src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/GraphQLMiddleware.cs

```diff
@@ -161,16 +157,29 @@ private async Task ExecuteAsync(HttpContext context, ISchemaFactory schemaServic
                     FieldImpact = _settings.FieldImpact
                 };
                 _.Listeners.Add(dataLoaderDocumentListener);
+                _.RequestServices = context.RequestServices;
             });
 
             context.Response.StatusCode = (int)(result.Errors == null || result.Errors.Count == 0
                 ? HttpStatusCode.OK
-                : result.Errors.Any(x => x.Code == RequiresPermissionValidationRule.ErrorCode)
+                : result.Errors.Any(x => (x is ValidationError && (x as ValidationError).Number == RequiresPermissionValidationRule.ErrorCode))
                     ? HttpStatusCode.Unauthorized
                     : HttpStatusCode.BadRequest);
 
             context.Response.ContentType = MediaTypeNames.Application.Json;
-            await documentWriter.WriteAsync(context.Response.Body, result);
+
+            await WriteAsync(context.Response.Body, result, documentWriter);
+        }
+
+        private async Task WriteAsync<T>(Stream stream2, T value, IDocumentWriter documentWriter, CancellationToken cancellationToken = default)
+        {
+            // needs to be always async, otherwise __schema request is not working, direct write into response does not work as serialize is using sync method inside
```

Comment:
Do you mean that somewhere in our grapqhl code with have a call to a `Serialize` method (NewtonSoft?) that is not async and this is triggered in `IDocumentWriter.WriteaAsync`.
If buffering is required, maybe use a smaller reused buffer instead of a single one as big as the payload, and pool it too? Another option might be to use `BufferedStream` between `documentWriter` and `context.Response.Body`.